src/trace_comparison.rs

```diff
@@ -130,10 +121,114 @@ pub fn generate_block_comparison(
+/// 1. Iterates through both lists simultaneously, comparing transactions by hash:
+///    - If hashes match, compares transactions directly.
+///    - If mismatch, searches for matching transaction in the other list.
+/// 2. Handles unmatched transactions by comparing with "Empty" and adjusting indices,
+///    with a preference for base transactions.
+fn compare_traces(base_traces: &[Value], native_traces: &[Value]) -> Value {
+    fn create_hash_map(traces: &[Value]) -> HashMap<String, usize> {
+        traces
+            .iter()
+            .enumerate()
+            .map(|(idx, trace)| {
+                (
+                    trace["transaction_hash"]
+                        .as_str()
+                        .expect("Transaction hash is not a string")
+                        .to_string(),
+                    idx,
+                )
+            })
+            .collect()
+    }
+
+    let mut compare_traces_result = Vec::new();
+    let mut base_idx = 0;
+    let mut native_idx = 0;
+
+    // To look up indices of transaction hashes, in the event that the ordering of transactions is different
+    let base_hash_map = create_hash_map(base_traces);
+    let native_hash_map = create_hash_map(native_traces);
+
+    while base_idx < base_traces.len() || native_idx < native_traces.len() {
+        if base_idx >= base_traces.len() {
+            // Only native transactions left
+            compare_traces_result.push(
+                ComparisonResult::new_different_native_only(native_traces[native_idx].clone())
+                    .into(),
+            );
+            native_idx += 1;
+        } else if native_idx >= native_traces.len() {
+            // Only base transactions left
+            compare_traces_result.push(
+                ComparisonResult::new_different_base_only(base_traces[base_idx].clone()).into(),
+            );
+            base_idx += 1;
+        } else {
+            // There are both base and native transactions left
+            let base_trace = &base_traces[base_idx];
+            let native_trace = &native_traces[native_idx];
+
+            let (base_tx_hash, native_tx_hash) = match (
+                base_trace["transaction_hash"].as_str(),
+                native_trace["transaction_hash"].as_str(),
+            ) {
+                (Some(base), Some(native)) => (base, native),
+                _ => panic!("Base trace and native trace are expected to be strings"),
+            };
+
+            if base_tx_hash == native_tx_hash {
+                let (base_trace, native_trace) =
+                    match (base_trace.as_object(), native_trace.as_object()) {
+                        (Some(base), Some(native)) => (base, native),
+                        _ => panic!("Base trace and native trace are expected to be objects"),
+                    };
+                compare_traces_result.push(generate_transaction_comparisons(
+                    base_trace.clone(),
```

Comment:
The clones here are the most non-optimal parts of the function. My alternative was to take ownership directly on the incoming array, but I felt that it affected readability slightly. Here is what it would look like using `mut Vec<Value>` instead of `&[Value]`:  3de05afcd9c38291bf4e1dd249973b9bcf57f44e

Reply:
I think this is good as is.

Reply:
I decided to use this commit instead, 3de05afcd9c38291bf4e1dd249973b9bcf57f44e.